From 7c443a142efa7372a9938f5aae6df9be6ba99ead Mon Sep 17 00:00:00 2001
From: Viktoria Maximova <viktoria.maksimova@intel.com>
Date: Wed, 4 Oct 2023 16:25:18 +0200
Subject: [PATCH 156/172] Reverse translation of arithmetic instructions for
 cooperative matrixes (#2166)

Implement translation via SPIR-V friendly calls, as:

the LLVM instructions are not capable to accept target extension types;
matrix arithmetic instructions require additional carry additional rules, which LLVM can not perform (for example while technically Add for vectors and (flattened) matrices is the same - yet for matrices we need to perform extra checks, also mul instruction is complitely different).
As for now some BE would need to recognize and define what to do with a call to __spirv_FMul(matrixA, matrixB). Better option is to map such SPIR-V to an intrinsic or define an appropriate type in LLVM (hence defining rules for GEP and other instructions) , but it's off the table now.
---
 lib/SPIRV/SPIRVReader.cpp                     |  9 +++++
 .../arithmetic_instructions.ll                | 33 ++++++++++++++-----
 2 files changed, 33 insertions(+), 9 deletions(-)

diff --git a/lib/SPIRV/SPIRVReader.cpp b/lib/SPIRV/SPIRVReader.cpp
index f7d00b59..a0693ba0 100644
--- a/lib/SPIRV/SPIRVReader.cpp
+++ b/lib/SPIRV/SPIRVReader.cpp
@@ -1083,6 +1083,9 @@ static void applyFPFastMathModeDecorations(const SPIRVValue *BV,
 Value *SPIRVToLLVM::transShiftLogicalBitwiseInst(SPIRVValue *BV, BasicBlock *BB,
                                                  Function *F) {
   SPIRVBinary *BBN = static_cast<SPIRVBinary *>(BV);
+  if (BV->getType()->isTypeCooperativeMatrixKHR()) {
+    return mapValue(BV, transSPIRVBuiltinFromInst(BBN, BB));
+  }
   Instruction::BinaryOps BO;
   auto OP = BBN->getOpCode();
   if (isLogicalOpCode(OP))
@@ -2320,6 +2323,9 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
       Builder.SetInsertPoint(BB);
     }
     SPIRVUnary *BC = static_cast<SPIRVUnary *>(BV);
+    if (BV->getType()->isTypeCooperativeMatrixKHR()) {
+      return mapValue(BV, transSPIRVBuiltinFromInst(BC, BB));
+    }
     auto Neg =
         Builder.CreateNeg(transValue(BC->getOperand(0), F, BB), BV->getName());
     if (auto *NegInst = dyn_cast<Instruction>(Neg)) {
@@ -2372,6 +2378,9 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
 
   case OpFNegate: {
     SPIRVUnary *BC = static_cast<SPIRVUnary *>(BV);
+    if (BV->getType()->isTypeCooperativeMatrixKHR()) {
+      return mapValue(BV, transSPIRVBuiltinFromInst(BC, BB));
+    }
     auto Neg = UnaryOperator::CreateFNeg(transValue(BC->getOperand(0), F, BB),
                                          BV->getName(), BB);
     applyFPFastMathModeDecorations(BV, Neg);
diff --git a/test/transcoding/SPV_KHR_cooperative_matrix/arithmetic_instructions.ll b/test/transcoding/SPV_KHR_cooperative_matrix/arithmetic_instructions.ll
index 82308f42..1758d491 100644
--- a/test/transcoding/SPV_KHR_cooperative_matrix/arithmetic_instructions.ll
+++ b/test/transcoding/SPV_KHR_cooperative_matrix/arithmetic_instructions.ll
@@ -6,10 +6,9 @@
 ; TODO: Validation is disabled till the moment the tools in CI are updated (passes locally)
 ; R/UN: spirv-val %t.spv
 
-; TODO: come up with an approach and implement reverse translation
-; R/UN: llvm-spirv -r --spirv-target-env=SPV-IR %t.spv -o %t.rev.bc
-; R/UN: llvm-dis %t.rev.bc
-; R/UN: FileCheck < %t.rev.ll %s --check-prefix=CHECK-LLVM
+; RUN: llvm-spirv -r --spirv-target-env=SPV-IR %t.spv -o %t.rev.bc
+; RUN: llvm-dis %t.rev.bc
+; RUN: FileCheck < %t.rev.ll %s --check-prefix=CHECK-LLVM
 
 ; CHECK-SPIRV: TypeInt [[#TypeInt:]] 32 0
 ; CHECK-SPIRV: TypeCooperativeMatrixKHR [[#MatrixTypeInt:]] [[#TypeInt]]
@@ -24,6 +23,8 @@ target triple = "spir-unknown-unknown"
 
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixIn:]] [[#]] {{$}}
 ; CHECK-SPIRV: SNegate [[#MatrixTypeInt]] [[#]] [[#MatrixIn]]
+; CHECK-LLVM: %1 = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z26__spirv_CompositeConstructi(i32 0)
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z15__spirv_SNegatePU3AS143__spirv_CooperativeMatrixKHR__int_3_12_12_3(%spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %1)
 define spir_kernel void @testSNegate(i32 %a) #0 !kernel_arg_addr_space !10 !kernel_arg_access_qual !11 !kernel_arg_type !12 !kernel_arg_type_qual !9 !kernel_arg_base_type !12 {
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
   %call = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z15__spirv_SNegate(%spirv.CooperativeMatrixKHR._int_3_12_12_3* %1)
@@ -32,7 +33,8 @@ define spir_kernel void @testSNegate(i32 %a) #0 !kernel_arg_addr_space !10 !kern
 
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixIn:]] [[#]] {{$}}
 ; CHECK-SPIRV: FNegate [[#MatrixTypeFloat]] [[#]] [[#MatrixIn]]
-; CHECK-LLVM: fneg
+; CHECK-LLVM: %0 = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z26__spirv_CompositeConstructf(float 0.000000e+00)
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z15__spirv_FNegatePU3AS145__spirv_CooperativeMatrixKHR__float_3_12_12_3(%spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %0)
 define spir_kernel void @testFNeg(float %a) local_unnamed_addr #0 !kernel_arg_addr_space !2 !kernel_arg_access_qual !3 !kernel_arg_type !4 !kernel_arg_base_type !4 !kernel_arg_type_qual !9 {
 entry:
   %0 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
@@ -43,6 +45,9 @@ entry:
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: IAdd [[#MatrixTypeInt]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %1 = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z26__spirv_CompositeConstructi(i32 0)
+; CHECK-LLVM: %2 = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z26__spirv_CompositeConstructi(i32 0)
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z12__spirv_IAddPU3AS143__spirv_CooperativeMatrixKHR__int_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %1, %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %2)
 define spir_kernel void @testIAdd(i32 %a, i32 %b) #0 !kernel_arg_addr_space !4 !kernel_arg_access_qual !5 !kernel_arg_type !6 !kernel_arg_type_qual !7 !kernel_arg_base_type !6 {
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
   %2 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
@@ -53,6 +58,7 @@ define spir_kernel void @testIAdd(i32 %a, i32 %b) #0 !kernel_arg_addr_space !4 !
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: ISub [[#MatrixTypeInt]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z12__spirv_ISubPU3AS143__spirv_CooperativeMatrixKHR__int_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %1, %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %2)
 define spir_kernel void @testISub(i32 %a, i32 %b) #0 !kernel_arg_addr_space !4 !kernel_arg_access_qual !5 !kernel_arg_type !6 !kernel_arg_type_qual !7 !kernel_arg_base_type !6 {
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
   %2 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
@@ -63,6 +69,7 @@ define spir_kernel void @testISub(i32 %a, i32 %b) #0 !kernel_arg_addr_space !4 !
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: IMul [[#MatrixTypeInt]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z12__spirv_IMulPU3AS143__spirv_CooperativeMatrixKHR__int_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %1, %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %2)
 define spir_kernel void @testIMul(i32 %a, i32 %b) #0 !kernel_arg_addr_space !4 !kernel_arg_access_qual !5 !kernel_arg_type !6 !kernel_arg_type_qual !7 !kernel_arg_base_type !6 {
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
   %2 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
@@ -73,6 +80,7 @@ define spir_kernel void @testIMul(i32 %a, i32 %b) #0 !kernel_arg_addr_space !4 !
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: SDiv [[#MatrixTypeInt]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z12__spirv_SDivPU3AS143__spirv_CooperativeMatrixKHR__int_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %1, %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %2)
 define void @testSDiv(i32 %a, i32 %b) {
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
   %2 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
@@ -83,6 +91,7 @@ define void @testSDiv(i32 %a, i32 %b) {
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeInt]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: UDiv [[#MatrixTypeInt]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* @_Z12__spirv_UDivPU3AS143__spirv_CooperativeMatrixKHR__int_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %1, %spirv.CooperativeMatrixKHR._int_3_12_12_3 addrspace(1)* %2)
 define void @testUDiv(i32 %a, i32 %b) {
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
   %2 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._int_3_12_12_3* @_Z26__spirv_CompositeConstructInt32(i32 0)
@@ -94,44 +103,50 @@ define void @testUDiv(i32 %a, i32 %b) {
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: FAdd [[#MatrixTypeFloat]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %0 = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z26__spirv_CompositeConstructf(float 0.000000e+00)
+; CHECK-LLVM: %1 = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z26__spirv_CompositeConstructf(float 0.000000e+00)
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z12__spirv_FAddPU3AS145__spirv_CooperativeMatrixKHR__float_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %1)
 define spir_kernel void @testFAdd(float %a, float %b) local_unnamed_addr #0 !kernel_arg_addr_space !2 !kernel_arg_access_qual !3 !kernel_arg_type !4 !kernel_arg_base_type !4 !kernel_arg_type_qual !5 {
 entry:
   %0 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
-  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FAdd(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1) 
+  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FAdd(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1)
   ret void
 }
 
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: FSub [[#MatrixTypeFloat]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z12__spirv_FSubPU3AS145__spirv_CooperativeMatrixKHR__float_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %1)
 define spir_kernel void @testFSub(float %a, float %b) local_unnamed_addr #0 !kernel_arg_addr_space !2 !kernel_arg_access_qual !3 !kernel_arg_type !4 !kernel_arg_base_type !4 !kernel_arg_type_qual !5 {
 entry:
   %0 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
-  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FSub(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1) 
+  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FSub(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1)
   ret void
 }
 
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: FMul [[#MatrixTypeFloat]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z12__spirv_FMulPU3AS145__spirv_CooperativeMatrixKHR__float_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %1)
 define spir_kernel void @testFMul(float %a, float %b) local_unnamed_addr #0 !kernel_arg_addr_space !2 !kernel_arg_access_qual !3 !kernel_arg_type !4 !kernel_arg_base_type !4 !kernel_arg_type_qual !5 {
 entry:
   %0 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
-  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FMul(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1) 
+  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FMul(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1)
   ret void
 }
 
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixA:]] [[#]] {{$}}
 ; CHECK-SPIRV: CompositeConstruct [[#MatrixTypeFloat]] [[#MatrixB:]] [[#]] {{$}}
 ; CHECK-SPIRV: FDiv [[#MatrixTypeFloat]] [[#]] [[#MatrixA]] [[#MatrixB]]
+; CHECK-LLVM: %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* @_Z12__spirv_FDivPU3AS145__spirv_CooperativeMatrixKHR__float_3_12_12_3S1_(%spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3 addrspace(1)* %1)
 define spir_kernel void @testFDiv(float %a, float %b) local_unnamed_addr #0 !kernel_arg_addr_space !2 !kernel_arg_access_qual !3 !kernel_arg_type !4 !kernel_arg_base_type !4 !kernel_arg_type_qual !5 {
 entry:
   %0 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
   %1 = tail call spir_func noundef %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z26__spirv_CompositeConstructFloat(float 0.000000e+00)
-  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FDiv(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1) 
+  %call = call spir_func %spirv.CooperativeMatrixKHR._float_3_12_12_3* @_Z12__spirv_FDiv(%spirv.CooperativeMatrixKHR._float_3_12_12_3* %0, %spirv.CooperativeMatrixKHR._float_3_12_12_3* %1)
   ret void
 }
 
-- 
2.20.1

